app/data_model/database.py

```diff
@@ -20,19 +20,24 @@ class QuestionnaireState(base):
     __tablename__ = 'questionnaire_state'
     user_id = Column('userid', String, primary_key=True)
     state = Column('questionnaire_data', String)
+    _data = None
 
     def __init__(self, user_id, data):
         self.user_id = user_id
+        self._data = data
         self.state = json.dumps(data)
 
     def set_data(self, data):
         logger.debug("setting data for questionnaire state")
-        self.state = json.dumps(data)
+        self._data = data
+        self.state = json.dumps(self._data)
 
     def get_data(self):
-        data = json.loads(self.state)
+        if not self._data:
```

Comment:
What happens if `self.state` has changed since the last call to `get_data`?

Reply:
i.e. can `self.state` be made private?

Reply:
state can't be private otherwise sqlalchemy doesn't persist it.